File: src/main/java/aurocosh/divinefavor/common/lib/EnumIndexer.java
package aurocosh.divinefavor.common.lib;

import aurocosh.divinefavor.common.util.UtilArray;

public class EnumIndexer<T extends Enum<T>> {
    private final T[] values;

    public EnumIndexer(T[] values) {
        this.values = values;
    }

    public T[] getValues() {
        return values;
    }

    public T get(int index) {
        return values[clampIndex(index)];
    }

    public int clampIndex(int index) {
        return UtilArray.clampIndex(values, index);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/lib/EnumIndexer.kt
package aurocosh.divinefavor.common.lib

import aurocosh.divinefavor.common.lib.extensions.getSafe

open class EnumIndexer<T : Enum<T>>(val values: Array<T>) {
    operator fun get(index: Int) = values.getSafe(index)
    fun getMaxOrdinal() = values.size - 1
}
